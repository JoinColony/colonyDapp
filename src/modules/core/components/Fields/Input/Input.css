@import 'variables.css';

.main {
  composes: main from '../Field/Field.css';
}

.stateActive {
  composes: stateActive from '../Field/Field.css';
}

.input {
  composes: input from '../Field/Field.css';
  display: flex;
  align-items: center;
  padding: 0 8px;
  height: 32px;
  width: 100%;
  border: 1px solid var(--temp-grey-5);
  border-radius: 2px;
  background-color: var(--colony-white);
  font-size: var(--size-normal);
  line-height: 32px;
  color: var(--text);
  letter-spacing: 0.25px;
  transition: border-color 0.1s ease-in;
}

.delegate {
  flex: 1 1 0;
  height: 100%;
  width: 100%;
  vertical-align: baseline;
  border: none;
  background-color: transparent;
  font-size: var(--size-normal);
  color: inherit;
  outline: none;

  &::placeholder {
    font-size: var(--size-normal);
    font-style: italic;
    letter-spacing: var(--spacing-normal);
  }
}

.themeFat {
  composes: main;
}

.themeFat .input {
  height: 45px;
  font-size: var(--size-medium);
  line-height: 45px;

  &::placeholder {
    font-weight: var(--weight-semi);
  }
}

.themeFat .delegate {
  font-size: var(--size-medium);

  &::placeholder {
    font-weight: var(--weight-semi);
  }
}

.themeUnderlined .input {
  display: inline-block;
  height: 45px;
  border: 0;
  border-bottom: 1px solid var(--colony-black);
  border-radius: 0;
  background-color: transparent;
  font-size: var(--size-large);
  text-align: right;
  line-height: 45px;
  color: var(--colony-black);
}

.themeUnderlined .input:focus {
  border-color: var(--primary);
}

.themeUnderlined[aria-invalid="true"] .input {
  /* danger is more important than primary */
  border-color: var(--danger) !important;
}

.directionHorizontal {
  display: flex;
  align-items: center;
  position: relative;
}

.alignRight .input {
  text-align: right;
}

.colorSchemaDark .input {
  border: 1px solid var(--grey-4);
  background-color: var(--temp-grey-2);
  color: var(--temp-grey-0);
  box-shadow: inset 0 5px 5px 0 rgba(0, 0, 0, 0.2);

  &::placeholder {
    color: var(--grey-4);
  }

  & svg {
    fill: var(--grey-4);
    stroke: var(--grey-4);
  }
}

.colorSchemaTransparent .input {
  border: 1px solid var(--grey-5);
  background-color: transparent;
  color: inherit;
  box-shadow: none;
}

.error {
  composes: errorHorizontal from '../Field/Field.css';
}

.controls {
  display: flex;
  align-items: center;
  color: inherit;
}

.controls svg {
  fill: var(--grey-5);
  stroke: var(--grey-5);
}
