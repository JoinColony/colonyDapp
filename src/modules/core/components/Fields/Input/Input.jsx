/* @flow */

import React, { Component } from 'react';
import Cleave from 'cleave.js/react';

import { getMainClasses } from '$utils/css';

import styles from './Input.css';

import type { CleaveHTMLInputElement, CleaveOptions, FieldComponentProps } from '../flowTypes';

import InputLabel from '../InputLabel';

/**
 * An input component (to be used as a redux form field) that is basically a standard (text|email|number) Input component.
 * Can receive formatting options leveraging cleave.js
 *
 * @method Input
 *
 * Can receive all the Field properties plus:
 *
 * @param {object} format Cleave.js formatting options
 */

type CustomProps = {
  formattingOptions?: CleaveOptions,
  disabled?: boolean,
};

type Props = FieldComponentProps<CustomProps>;

class Input extends Component<Props> {
  // TODO: exact
  handleBlur: Function;
  handleChange: Function;
  handleFocus: Function;
  handleRef: Function;
  inputElm: HTMLInputElement;
  static displayName = 'core.Fields.Input';
  static defaultProps = {
    appearance: {},
  };
  constructor(props: Props) {
    super(props);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.handleRef = this.handleRef.bind(this);
  }
  handleBlur(evt: SyntheticEvent<CleaveHTMLInputElement>) {
    return this.props.input.onBlur(evt.currentTarget.rawValue);
  }
  handleChange(evt: SyntheticEvent<CleaveHTMLInputElement>) {
    return this.props.input.onChange(evt.currentTarget.rawValue);
  }
  // We're using a simple object here because redux-form is using its
  // own event definition which doesn't make any sense
  handleFocus(evt: Object) {
    const { onFocus } = this.props.input;
    const { length } = evt.currentTarget.value;
    setTimeout(() => {
      if (!this.inputElm) return;
      this.inputElm.selectionStart = length;
      this.inputElm.selectionEnd = length;
    });
    return onFocus(evt);
  }
  handleRef(elm: HTMLInputElement) {
    this.inputElm = elm;
  }
  render() {
    const {
      appearance,
      elementOnly,
      error,
      hasError,
      help,
      input,
      inputProps,
      passthroughProps: { formattingOptions, disabled, ...props },
      label,
      meta: { active },
    } = this.props;
    return (
      <div
        className={getMainClasses(appearance, styles, { active })}
        aria-invalid={hasError}
        aria-disabled={disabled}
      >
        {!elementOnly && label ?
          <InputLabel id={inputProps.id} label={label} error={hasError && error} help={help} appearance={appearance} />
          :
          null
        }
        { formattingOptions ?
          <Cleave
            {...props}
            {...inputProps}
            {...input}
            className={styles.input}
            options={formattingOptions}
            htmlRef={this.handleRef}
            onChange={this.handleChange}
            onBlur={this.handleBlur}
            onFocus={this.handleFocus}
            value={active ? input.value : (input.value || inputProps.placeholder)}
          />
          :
            <input
              {...props}
              {...inputProps}
              {...input}
              className={styles.input}
            />
        }
        {appearance && appearance.direction === 'horizontal' ?
          <span className={styles.error}>{hasError && error}</span>
          :
          null
        }
      </div>
    );
  }
}

export default Input;
