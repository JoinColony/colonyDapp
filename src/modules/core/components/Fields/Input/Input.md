### Input text

```js
<Input
  elementOnly
  name="input"
  placeholder="I'm an input"
  label="Label"
  connect={false}
/>
```

### Input text with error

```js
<Input
  elementOnly
  name="inputerror"
  placeholder="Hover over me to see error"
  label="Label"
  form={{ errors: { inputerror: 'Wrong!' }, touched: { inputerror: true }}}
  connect={false}
/>
```

### Input text with label

```js
<Input
  name="inputWithlabel"
  placeholder="I have a label"
  label="Your details"
  connect={false}
/>
```

### Input text aligned right

```js
<Input
  appearance={{ align: 'right' }}
  name="inputlabelright"
  placeholder="I'm on the right side!"
  label="You came to the right place"
  connect={false}
/>
```

### Input text with label and error

```js
<Input
  name="inputlabelerror"
  placeholder="I'm wrong as well!"
  label="Label"
  form={{ errors: { inputlabelerror: 'Wrong!' }, touched: { inputlabelerror: true }}}
  connect={false}
/>
```

### Input text with label and status

```js
<Input
  name="inputlabelstatus"
  placeholder="I have a custom status!"
  label="Look!"
  status="This is all good now"
  connect={false}
/>
```

### Input text with label and extra node

```js
<Input
  name="inputlabelextra"
  placeholder="I have an extra node!"
  label="Look right ->"
  extra={<a href="#">I'm an extra node!</a>}
  connect={false}
/>
```

### Input text with horizontal label

```js
<Input
  appearance={{ direction: 'horizontal' }}
  name="inputhorizontal"
  placeholder="I'm horizontal!"
  label="Label"
  connect={false}
/>
```

### Input text with horizontal label and error

```js
<Input
  appearance={{ direction: 'horizontal' }}
  name="inputhorizontalerror"
  placeholder="I'm horizontal and wrong!"
  label="Label"
  form={{ errors: { inputhorizontalerror: 'Wrong!' }, touched: { inputhorizontalerror: true }}}
  connect={false}
/>
```

### Input with dark color schema

Looks great on dark backgrounds. This should also work for the fat theme.

```js
<Input
  appearance={{ colorSchema: 'dark' }}
  name="inputdark"
  placeholder="I'm dark!"
  label="Dark label"
  connect={false}
/>
```

### Input with transparent color schema

Uses the background color of its background (same for text). This should also work for the fat theme.

```js
<Input
  appearance={{ colorSchema: 'transparent' }}
  name="inputtransparent"
  placeholder="I'm a placeholder"
  label="Cool label"
  connect={false}
/>
```

### Fat input field

```js
<Input
  appearance={{ theme: 'fat' }}
  name="inputfat"
  placeholder="I'm fat"
  label="Fat label"
  connect={false}
/>
```

### Underlined theme

```js
<Input
  appearance={{ theme: 'underlined', direction: 'horizontal' }}
  name="inputunderlined"
  placeholder="I'm underlined"
  label="A label"
  connect={false}
/>
```

### Underlined theme error

```js
<Input
  appearance={{ theme: 'underlined', direction: 'horizontal' }}
  name="inputunderlinederror"
  placeholder="I'm underlined"
  label="A label"
  form={{ errors: { inputunderlinederror: 'A meaningful error' }, touched: { inputunderlinederror: true }}}
  connect={false}
/>
```

### Minimal theme

```js
<Input
  appearance={{ theme: 'minimal' }}
  name="inputminimal"
  placeholder="I'm minimal"
  label="Minimal label"
  connect={false}
/>
```

### Underline Dotted theme

```js
<Input
  appearance={{ theme: 'dotted', colorSchema: 'grey' }}
  name="inputdotted"
  placeholder="I have funny dots"
  label="Dotted label"
  connect={false}
/>
```

### Input that displays extension

```js
<Input
  name="ENSname"
  appearance={{ theme: 'fat' }}
  label="Tell me your name"
  extensionString=".joincolony.eth"
  connect={false}
/>
```

### Formatting using [Cleave.js](https://nosir.github.io/cleave.js/)

```js
<Input
  name="inputcleave"
  placeholder="Put in a big number"
  label="Number formatted"
  formattingOptions={{ numeral: true, delimiter: ',' }}
  connect={false}
/>
```

### A form embedded in Formik

```js
const { Formik } = require('formik');
<Formik
  initialValues={{
    forminput1: '',
    forminput2: '',
  }}
  onSubmit={values => console.log(values)}
  render={
    ({ handleSubmit }) => (
      <form onSubmit={handleSubmit}>
        <Input
          name="forminput1"
          label="Type a string"
          placeholder="Type here"
        />
        <Input
          name="forminput2"
          label="Type a number"
          formattingOptions={{ numeral: true, delimiter: ',' }}
          placeholder="And here"
        />
        <Button appearance={{ theme: 'primary' }} type="submit">And press me</Button>
      </form>
    )
  }
/>
```
