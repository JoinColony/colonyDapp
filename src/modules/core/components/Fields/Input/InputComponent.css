.main {
  padding: 0 16px;
  height: 32px;
  width: 100%;
  border: 1px solid var(--temp-grey-5);
  border-radius: var(--radius-tiny);
  background-color: var(--colony-white);
  font-size: var(--size-normal);
  line-height: 32px;
  color: var(--text);
  outline: none;
  letter-spacing: var(--spacing-normal);
  transition: border-color 0.1s ease-in;

  &:focus {
    border-color: var(--primary);
  }

  &::placeholder {
    font-size: var(--size-normal);
    font-style: italic;
    letter-spacing: var(--spacing-normal);
  }
}

.main[aria-invalid="true"] {
  border-color: var(--danger);
}

.themeFat {
  composes: main;
  height: 45px;
  font-size: var(--size-medium);
  line-height: 45px;

  &::placeholder {
    font-weight: var(--weight-semi);
  }
}

.themeUnderlined {
  display: inline-block;
  height: 45px;
  border: 0;
  border-bottom: 1px solid var(--colony-black);
  border-radius: 0 !important;
  background-color: transparent;
  font-size: var(--size-large);
  text-align: right;
  line-height: 45px;
  color: var(--colony-black);
  outline: none;
}

.themeUnderlined:focus {
  border-color: var(--primary);
}

.themeUnderlined[aria-invalid="true"] {
  border-color: var(--danger);
}

.themeMinimal {
  display: block;
  padding: 5px 0;
  width: 100%;
  border: 0;
  border-radius: 0;
  border-bottom: 1px solid var(--grey-4);
  background-color: transparent;
  font-size: var(--size-medium);
  font-weight: var(--weight-medium);
  color: var(--grey-dark);
  outline: none;

  &:focus {
    border-color: var(--primary);
  }
}

.themeMinimal::placeholder {
  font-style: italic;
  color: var(--grey-4);
}

.themeMinimal[aria-invalid="true"] {
  border-color: var(--danger);
}

.themeDotted {
  display: block;
  padding: 5px 0;
  border: 0;
  border-radius: 0;
  border-bottom: 1px solid var(--grey-5);
  border-style: dotted;
  background-color: transparent;
  font-size: var(--size-medium);
  font-weight: var(--weight-medium);
  color: var(--temp-grey-4);
  outline: none;
}

.themeDotted::placeholder {
  color: var(--text);
  opacity: 0.5;
}

.alignRight {
  text-align: right;
}

.sizeSmall {
  font-size: var(--size-small);
}

.colorSchemaGrey {
  border: 1px solid rgb(200, 214, 245);
  background-color: rgb(249, 250, 250);
  color: rgb(60, 68, 77);
  box-shadow: inset 0px 2px 4px rgba(0, 0, 0, 0.07);
}

.colorSchemaDark {
  border: 1px solid var(--grey-4);
  background-color: var(--temp-grey-2);
  color: var(--temp-grey-0);
  box-shadow: inset 0 5px 5px 0 rgba(0, 0, 0, 0.2);

  &::placeholder {
    color: var(--grey-4);
  }

  & svg {
    fill: var(--grey-4);
    stroke: var(--grey-4);
  }
}

.colorSchemaTransparent {
  border: 1px solid var(--grey-5);
  background-color: transparent;
  color: inherit;
  box-shadow: none;
}
