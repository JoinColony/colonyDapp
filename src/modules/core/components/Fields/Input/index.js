/* @flow */

export { default } from './Input';
