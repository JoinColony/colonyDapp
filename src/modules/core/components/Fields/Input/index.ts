export { default } from './Input';
export { default as InputComponent, MaxButtonParams } from './InputComponent';

export interface InputComponentAppearance {
  theme?: 'fat' | 'underlined' | 'minimal' | 'dotted';
  align?: 'right';
  direction?: 'horizontal';
  colorSchema?: 'dark' | 'grey' | 'transparent';
  size?: 'small';
}
