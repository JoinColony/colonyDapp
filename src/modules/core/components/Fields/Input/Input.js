/* @flow */

import InputComponent from './Input.jsx';
import { asField } from '../Field';

export default asField()(InputComponent);
