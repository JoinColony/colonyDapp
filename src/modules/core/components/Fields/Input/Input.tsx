import React, { ReactNode } from 'react';
import cx from 'classnames';
import { CleaveOptions } from 'cleave.js/options';
import { MessageDescriptor } from 'react-intl';

import asField from '../asField';
import InputLabel from '../InputLabel';
import InputStatus from '../InputStatus';
import { FieldEnhancedProps } from '../types';
import InputComponent, { Props as InputComponentProps } from './InputComponent';

import styles from './Input.css';

interface Appearance {
  theme?: 'fat' | 'underlined' | 'minimal' | 'dotted';
  align?: 'right';
  direction?: 'horizontal';
  colorSchema?: 'dark' | 'grey' | 'transparent';
  helpAlign?: 'right';
  size?: 'small';
}

export interface Props {
  /** Appearance object */
  appearance?: Appearance;

  /** Add extension of input to the right of it, i.e. for ENS name */
  extensionString?: string | MessageDescriptor;

  /** Extra node to render on the top right in the label */
  extra?: ReactNode;

  /** Options for cleave.js formatting (see [this list](https://github.com/nosir/cleave.js/blob/master/doc/options.md)) */
  formattingOptions?: CleaveOptions;

  /** Pass a ref to the `<input>` element */
  innerRef?: (ref: HTMLElement | null) => void;
}

const Input = ({
  appearance = {},
  elementOnly,
  extensionString,
  extra,
  formattingOptions,
  formatIntl,
  help,
  innerRef,
  $id,
  label,
  name,
  onChange,
  status,
  $value,
  $error,
  /* eslint-disable @typescript-eslint/no-unused-vars */
  connect,
  setValue,
  setError,
  isSubmitting,
  $touched,
  /* eslint-enable @typescript-eslint/no-unused-vars */
  ...props
}: Props & FieldEnhancedProps<string, InputComponentProps>) => {
  const inputProps: InputComponentProps = {
    appearance,
    'aria-invalid': $error ? true : undefined,
    formattingOptions,
    id: $id,
    innerRef,
    name,
    onChange,
    value: undefined as (string | undefined),
    ...props,
  };

  // If there is an onChange handler, make it a controlled input
  if (onChange) {
    inputProps.value = $value;
  }

  const extensionStringText: string | undefined =
    !extensionString || typeof extensionString === 'string'
      ? extensionString
      : formatIntl(extensionString);

  if (elementOnly) {
    return <InputComponent {...inputProps} />;
  }
  const containerClasses = cx(styles.container, {
    [styles.containerHorizontal]: appearance.direction === 'horizontal',
  });
  return (
    <div className={containerClasses}>
      <InputLabel
        appearance={appearance}
        inputId={$id}
        label={label}
        help={help}
        extra={extra}
      />
      <div className={styles.extensionContainer}>
        <InputComponent {...inputProps} />
        {extensionStringText && (
          <div className={styles.extension}>{extensionStringText}</div>
        )}
      </div>
      <InputStatus appearance={appearance} status={status} error={$error} />
    </div>
  );
};

Input.displayName = 'Input';

export default asField<Props, string, InputComponentProps>({
  initialValue: '',
})(Input);
