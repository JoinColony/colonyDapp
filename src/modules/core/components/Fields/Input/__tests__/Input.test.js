/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl } from '$utils/test';

import Cleave from 'cleave.js/react';
import Input from '../Input.jsx';
import InputLabel from '../../InputLabel';

jest.mock('../../InputLabel');

describe('Input', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl((
      <Input
        value=""
        type="text"
        placeholder="Name"
        input={{}}
        inputProps={{ id: 'blah' }}
        passthroughProps={{ formattingOptions: { initValue: 'none' } }}
        meta={{ active: false }}
        title="Invitee's name"
        label="Invitee's name"
        elementOnly={false}
        hasError={false}
        data-wd-hook="invite-name"
      />
    ));
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('If elementOnly and no formatting options, InputLabel and Cleave are not rendered', () => {
    const wrapper = shallowWithIntl((
      <Input
        name="name"
        value=""
        type="text"
        placeholder="Name"
        input={{}}
        inputProps={{}}
        passthroughProps={{}}
        meta={{ active: false }}
        title="Invitee's name"
        label="Invitee's name"
        elementOnly
      />
    ));
    expect(wrapper.find(Cleave)).not.toBePresent();
    expect(wrapper.find(InputLabel)).not.toBePresent();
  });
});
