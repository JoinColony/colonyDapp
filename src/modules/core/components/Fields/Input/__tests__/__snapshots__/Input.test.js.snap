// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Renders initial component 1`] = `
<div
  aria-invalid={false}
  className=""
>
  <InjectIntl(core.InputLabel)
    appearance={Object {}}
    error={false}
    id="blah"
    label="Invitee's name"
  />
  <Component
    htmlRef={[Function]}
    id="blah"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    options={
      Object {
        "initValue": "none",
      }
    }
  />
</div>
`;
